### PR TITLE
Add sparse inference & training benchmark

### PR DESCRIPTION
**Pull request by kalyc**
### Summary
Add inference benchmark for sparse

### PR Overview

- [ ] This PR requires new unit tests [y/n] (make sure tests are included)
- [ ] This PR requires to update the documentation [y/n] (make sure the docs are up-to-date)
- [ ] This PR is backwards compatible [y/n]
- [ ] This PR changes the current API [y/n]

### CODE REVIEW

File: benchmark/sparse/linear_regression/inference_benchmark.md [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+| :--------------- | :----------------------------------------------------------- |
+| Keras            | v2.2.2                                                      |
+| TensorFlow       | v1.11.0                                                     |
+| MXNet-mkl         | v1.3.0   
+
+###### Using 10 epochs
+#### CPU
+##### Speed
+###### Note speed calculated using an average of 5 runs
+| Instance Type | GPUs  | Batch Size  | Keras-MXNet (Time/Epoch) | Keras-TensorFlow (Time/Epoch)  |
```
Comment (roywei) [inline benchmark/sparse/linear_regression/inference_benchmark.md]: Why is inference benchmarked as time/epoch? Only training (model.fit) ran several epochs. For inference, you did model.predict(data), the time measured should be the time to run inference on total data. So the smaller the number better performance. We can show in images per second using `time/len(data)`

File: benchmark/sparse/linear_regression/keras_sparse_model.py [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+        model = multi_gpu_model(model, gpus=gpus)
+
+    model.compile(loss='mse', optimizer=sgd, metrics=['accuracy'])
+
+    model.fit(train_data,
+              train_label,
+              epochs=epochs,
+              batch_size=batch_size, verbose=1)
+
+    start = time.time()
```
Comment (roywei) [inline benchmark/sparse/linear_regression/keras_sparse_model.py]: Let's add an argument `is_train`, or `benchmark_mode`, so users can configure what to benchmark. Time fit for training and only do predict and time for inference.

File: benchmark/sparse/linear_regression/inference_benchmark.md [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+| P3.8XLarge |   1  | 256 | 973.1 | 16135.5
+| P3.8XLarge |   1  | 512 | 994.1 | 16135.5
+| P3.8XLarge |   1  | 1024 | 987.7 | 16135.5
+
+##### Multi-GPU
+##### Speed
+###### Benchmark results on multi GPU calculated using an average of 5 runs
+| Instance Type | GPUs  | Batch Size | Keras-MXNet (Time/Epoch) | Keras-TensorFlow (Time/Epoch)  |
+|-----|-----|-----|-----|-----|
+| P3.8XLarge |   2  | 512 | 0.86 sec | Not supported
```
Comment (roywei) [inline benchmark/sparse/linear_regression/inference_benchmark.md]: We generally don't use multi-gpu for inference, and here the inference speed is actually slower than 1 GPU for batch 512 and 1024. Please verify or remove this section.

File: benchmark/sparse/linear_regression/inference_benchmark.md [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+|-----|-----|-----|-----|-----|
+| C5.8XLarge |   0  | 64  | 1630.8 | 1573.8 |
+| C5.8XLarge |   0  | 128 | 1574.7 | 1561.2 | 
+| C5.8XLarge |   0  | 256 | 1477.8 | 1501.4  |
+| C5.8XLarge |   0  | 512 | 1407.0| 1472.5 |
+| C5.8XLarge |   0  | 1024 | 1336.3 | 1466.8 |
+
+#### GPU
+### Configuration
+| Dataset          | Synthetic(Randomly generated)                                |
```
Comment (lupesko) [inline benchmark/sparse/linear_regression/inference_benchmark.md]: Same comment as above about the feature vector size

File: benchmark/sparse/linear_regression/training_benchmark.md [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+###### Note speed calculated using an average of 5 runs
+| Instance Type | GPUs  | Batch Size  | Keras-MXNet (Time/Epoch) | Keras-TensorFlow (Time/Epoch)  |
+|-----|-----|-----|-----|-----|
+| C5.8XLarge |   0  | 64  | 715.4 sec | 676.6 sec
+| C5.8XLarge |   0  | 128 | 346.2 sec | 341.1 sec 
+| C5.8XLarge |   0  | 256 | 168.9 sec | 165.5
+| C5.8XLarge |   0  | 512 | 89.3 sec | 83.8 sec 
+| C5.8XLarge |   0  | 1024 | 50.5 sec | 48.6 sec
+
+#### Memory consumed
```
Comment (lupesko) [inline benchmark/sparse/linear_regression/training_benchmark.md]: We should have memory in MB as well (like you have for inference)

File: benchmark/sparse/linear_regression/training_benchmark.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+##### Single GPU
+##### Speed
+###### Note speed calculated using an average of 5 runs
+| Instance Type | GPUs  | Batch Size  | Keras-MXNet (Time/Epoch) | Keras-TensorFlow (Time/Epoch)  |
+|-----|-----|-----|-----|-----|
+| P3.8XLarge |   1  | 64  | 62.3 sec | 59.7 sec
+| P3.8XLarge |   1  | 128 | 34.8 sec | 37.7 sec 
+| P3.8XLarge |   1  | 256 | 19.3 sec | 28.4 sec
+| P3.8XLarge |   1  | 512 | 19.3 sec | 28.4 sec
+| P3.8XLarge |   1  | 1024 | 19.3 sec | 28.4 sec
```
Comment (lupesko) [inline benchmark/sparse/linear_regression/training_benchmark.md]: Can we go to batch sizes higher than 1024?

File: benchmark/sparse/linear_regression/training_benchmark.md [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+
+##### Memory consumed
+| Instance Type | GPUs  | Batch Size | Keras-MXNet (%) | Keras-TensorFlow (%)  |
```
Comment (lupesko) [inline benchmark/sparse/linear_regression/training_benchmark.md]: Is this the CPU memory or GPU? (should be GPU).
Can you add a table with the absolute value in MB?

File: benchmark/sparse/linear_regression/inference_benchmark.md (hunk 3 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/inference_benchmark.md]: Removed

File: benchmark/sparse/linear_regression/inference_benchmark.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+### Results
+### Inference Benchmark
+### Configuration
+| Dataset          | Synthetic(Randomly generated)                                |
+| :--------------- | :----------------------------------------------------------- |
+| Keras            | v2.2.2                                                      |
+| TensorFlow       | v1.11.0                                                     |
+| MXNet-mkl         | v1.3.0   
+
+###### Using 10 epochs
```
[the remark this answers is not in the document]
Comment (kalyc) [reply benchmark/sparse/linear_regression/inference_benchmark.md]: Removed

File: benchmark/sparse/linear_regression/training_benchmark.md (hunk 5 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/training_benchmark.md]: Added

File: benchmark/sparse/linear_regression/training_benchmark.md (hunk 7 above)
Comment (kalyc) [reply benchmark/sparse/linear_regression/training_benchmark.md]: Updated table headers

**Comment by kalyc** [comment]
PR continued here - https://github.com/awslabs/keras-apache-mxnet/pull/197

Note - We have already pushed the benchmark scripts

**Pull request closed without merging**